rust/otap-dataflow/crates/otlp/src/batch_processor.rs

```diff
@@ -0,0 +1,1097 @@
+                let batch = self.get_batch(&metadata);
+                batch.push(data);
+                
+                // Check if we should emit this batch
+                if batch.len() >= self.config.send_batch_size as usize {
+                    let batch = std::mem::replace(batch, Vec::with_capacity(self.config.send_batch_size as usize));
+                    for item in batch {
+                        effect_handler.send_message(item).await?;
+                    }
+                    self.last_batch_time = Instant::now();
+                } else if self.config.timeout.as_secs() > 0 && self.last_batch_time.elapsed() >= self.config.timeout {
+                    let batch = std::mem::replace(batch, Vec::with_capacity(self.config.send_batch_size as usize));
+                    for item in batch {
+                        effect_handler.send_message(item).await?;
+                    }
+                    self.last_batch_time = Instant::now();
+                }
+                Ok(())
+            }
+            Message::Control(ctrl_msg) => {
+                // Log every ControlMsg variant as it is discovered
+                log::info!("[BatchProcessor] Received ControlMsg: {:?}", ctrl_msg);
+                match ctrl_msg {
+                    ControlMsg::TimerTick { .. } => {
+                        if !self.batches.is_empty() {
+                            for (_, batch) in self.batches.drain() {
+                                for item in batch {
+                                    effect_handler.send_message(item).await?;
+                                }
+                            }
+                        }
+                        Ok(())
+                    }
+                    ControlMsg::Shutdown { deadline: _, reason: _ } => {
+                        // Flush any remaining batch on shutdown
+                        if !self.batches.is_empty() {
+                            for (_, batch) in self.batches.drain() {
+                                for item in batch {
+                                    effect_handler.send_message(item).await?;
+                                }
+                            }
+                        }
+                        Ok(())
+                    }
+                    ControlMsg::Ack { id: _ } => {
+                        // Optionally handle ack logic here (e.g., mark message as acknowledged)
+                        Ok(())
+                    }
+                    ControlMsg::Nack { id: _, reason: _ } => {
+                        // Optionally handle nack logic here (e.g., log or retry)
+                        Ok(())
+                    }
+                    ControlMsg::Config { config: _ } => {
+                        // Optionally handle config update here (e.g., update batch_size)
+                        Ok(())
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These tests must adapted to the new test infrastructure recently introduced in the pipeline crate.